### PR TITLE
Add one-level list encoding support in parquet reader

### PR DESCRIPTION
**Pull request by PointKernel**
Closes https://github.com/rapidsai/cudf/issues/9240

This PR added the [one-level list encoding](https://github.com/apache/parquet-cpp/blob/master/src/parquet/schema.h#L43-L77) support in parquet reader. It also involved cleanups like removing the unused stream argument and fixing typos in docs/comments.

### CODE REVIEW

**Comment by PointKernel** [comment]
This PR can pass the unit test provided in https://github.com/rapidsai/cudf/issues/9240. I just want to make sure it covers all corner cases. @revans2 can you please test this PR with a larger file or other data types (like decimal, string, etc.)?

**Comment by revans2** [comment]
@res-life you found this bug is there additional testing that you want done?

**Review by devavret — COMMENTED** [review]
If you're looking for a way to add tests, you can either add the file provided in #9240 as binary to a gtest or add the file in python/cudf/cudf/tests/data and add a pytest.

File: cpp/src/io/parquet/parquet.hpp [hunk 1]
```diff
@@ -198,6 +198,12 @@ struct SchemaElement {
   //  };
   // }
   bool is_stub() const { return repetition_type == REPEATED && num_children == 1; }
+
+  // https://github.com/apache/parquet-cpp/blob/master/src/parquet/schema.h
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/parquet/parquet.hpp]: `pytest` added.

**Review by devavret — APPROVED** [review]

File: cpp/src/io/parquet/reader_impl.cu [hunk 2]
```diff
@@ -411,6 +411,9 @@ class aggregate_metadata {
     // walk upwards, skipping repeated fields
     while (schema_index > 0) {
       if (!pfm.schema[schema_index].is_stub()) { depth++; }
+      // schema of one-level encoding list doesn't contain nesting information, so we need to
+      // manually add an extra nesting level
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl.cu:415]: what a weird piece of specs

File: cpp/src/io/parquet/reader_impl.cu [hunk 3]
```diff
@@ -629,6 +635,26 @@ class aggregate_metadata {
...
+              schema_elem, strings_to_categorical, timestamp_type_id, strict_decimal_types);
+            auto const element_dtype =
+              element_type == type_id::DECIMAL32 || element_type == type_id::DECIMAL64
+                ? data_type{element_type, numeric::scale_type{-schema_elem.decimal_scale}}
+                : data_type{element_type};
+
+            column_buffer element_col(element_dtype, schema_elem.repetition_type == OPTIONAL);
+            // store the index of this element
+            nesting.push_back(static_cast<int>(output_col.children.size()));
+            // TODO: not sure if we should assign a name or leave it blank
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl.cu:652]: Is the included in the output schema?
In ORC the names of nested columns are generated as the index in the parent's list of children. Gives a uniform way to access nested columns of lists/maps/structs. I don't know enough about Parquet to understand if the same logic can apply here.

File: cpp/src/io/parquet/reader_impl.cu [hunk 4]
```diff
@@ -629,6 +635,26 @@ class aggregate_metadata {
...
+
+          // set up child output column for one-level encoding list
+          if (schema_elem.is_one_level_list()) {
+            // determine the element data type
+            auto const element_type = to_type_id(
+              schema_elem, strings_to_categorical, timestamp_type_id, strict_decimal_types);
+            auto const element_dtype =
+              element_type == type_id::DECIMAL32 || element_type == type_id::DECIMAL64
+                ? data_type{element_type, numeric::scale_type{-schema_elem.decimal_scale}}
+                : data_type{element_type};
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl.cu:647]: This block repeats now, might be good to add `type_id_to_dtype` (or to_data_type), that generates `data_type` from `type_id` and `SchemaElement`.
Also, maybe `data_type` should have a ctor that ignores scale if the type is not a fixed point type, but that's a wider topic.

**Review by vuule — COMMENTED** [review]
Looks good for the most part, thanks for taking care of this issue!

**Comment by res-life** [comment]
> @res-life you found this bug is there additional testing that you want done?
Yes, I'll update the test cases after this is merged.

File: cpp/src/io/parquet/reader_impl.cu (hunk 2 above)
Comment (PointKernel) [reply cpp/src/io/parquet/reader_impl.cu:415]: Agreed. It's a flat schema representing a nested data structure.

**Comment by PointKernel** [comment]
> > @res-life you found this bug is there additional testing that you want done?
> > Yes, I'll update the test cases after this is merged.

@res-life Just to be clear, the corresponding pytest has been added in this PR. Do you have other similar files to test before this PR gets merged?

**Comment by res-life** [comment]
@PointKernel  Please also check this:
[index-outof-bound.zip](https://github.com/rapidsai/cudf/files/7682437/index-outof-bound.zip)

**Comment by PointKernel** [comment]
rerun tests

**Review by vuule — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Comment by PointKernel** [comment]
@gpucibot merge

**Pull request merged**